liteeth/mac/sram.py

```diff
@@ -145,50 +195,99 @@ def __init__(self, dw, depth, nslots=2, endianness="big"):
         self.lengthbits = lengthbits
 
+        # TX packet timestamping, for applications such as IEEE 1588
+        # Precision Time Protocol
+        timestampbits   = 0 if timestamp_source is None else value_bits_sign(timestamp_source)[0]
+
+        # MAC Reader command channel
         self._start  = CSR()
         self._ready  = CSRStatus()
         self._level  = CSRStatus(log2_int(nslots) + 1)
         self._slot   = CSRStorage(slotbits,   reset_less=True)
         self._length = CSRStorage(lengthbits, reset_less=True)
 
+        # MAC Reader return channel
+        self._res_slot = CSRStatus(slotbits)
+        if timestamp_source is not None:
+            self._res_tx_timestamp = CSRStatus(timestampbits)
+
         self.submodules.ev = EventManager()
-        self.ev.done       = EventSourcePulse()
+        self.ev.done       = EventSourceLevel()
         self.ev.finalize()
 
         # # #
 
         # Command FIFO
-        fifo = stream.SyncFIFO([("slot", slotbits), ("length", lengthbits)], nslots)
-        self.submodules += fifo
+        cmd_fifo = stream.SyncFIFO([("slot", slotbits), ("length", lengthbits)], nslots)
+        self.submodules += cmd_fifo
+        self.comb += [
+            cmd_fifo.sink.valid.eq(self._start.re),
+            cmd_fifo.sink.slot.eq(self._slot.storage),
+            cmd_fifo.sink.length.eq(self._length.storage),
+            self._ready.status.eq(cmd_fifo.sink.ready),
+            self._level.status.eq(cmd_fifo.level)
+        ]
+
+        # Result FIFO (return channel)
+        if timestamp_source is not None:
+            res_fifo = stream.SyncFIFO([("slot", slotbits), ("tx_timestamp", timestampbits)], nslots)
+        else:
+            res_fifo = stream.SyncFIFO([("slot", slotbits)], nslots)
+
+        self.submodules += res_fifo
         self.comb += [
-            fifo.sink.valid.eq(self._start.re),
-            fifo.sink.slot.eq(self._slot.storage),
-            fifo.sink.length.eq(self._length.storage),
-            self._ready.status.eq(fifo.sink.ready),
-            self._level.status.eq(fifo.level)
+            res_fifo.source.ready.eq(self.ev.done.clear),
+            self.ev.done.trigger.eq(res_fifo.source.valid),
+            self._res_slot.status.eq(res_fifo.source.slot),
         ]
+        if timestamp_source is not None:
+            self.comb += [
+                self._res_tx_timestamp.status.eq(res_fifo.source.tx_timestamp),
+            ]
 
         # Length computation
         read_address = Signal(lengthbits)
         counter      = Signal(lengthbits)
 
+        # Store the timestamp of transmission start
+        if timestamp_source is not None:
+            tx_start_timestamp = Signal(timestampbits)
+
+            # Effectively build a latch, updating the timestamp only
+            # when starting the transmission
+            self.comb += [
+                tx_start_timestamp.eq(tx_start_timestamp)
+            ]
```

Comment:
Avoid multi-lines here.

Reply:
In fact, we should avoid latches, so you can probably remove this code and use NextValue(tx_start_timestamp, timestamp_source) in the FSM.